### PR TITLE
Rust language added patterns

### PR DESCRIPTION
**Pull request by guyacosta**
#235 for added language support for Rust.  Base branch is main against which to compare.

### CODE REVIEW

File: AppInspector/rules/default/cryptography/key_derivation.json [hunk 1]
```diff
@@ -42,7 +49,7 @@
     "description": "Cryptography: Key Derivation (PBKDF2)",
     "tags": [
       "Cryptography.KeyDerivation.PBKDF2",
-      "Cryptography.HashAlgorithm.PBKDF1"
+      "Cryptography.HashAlgorithm.PBKDF2"
```
Comment (scovetta) [inline AppInspector/rules/default/cryptography/key_derivation.json:52]: I think you can remove this line -- PBKDF2 is only a KDF. But PBKDF1 is also a thing (https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.passwordderivebytes?view=netcore-3.1 and https://github.com/tomyun/crypto-js/blob/d5ac13743fc6af51c2d4bb77986ba045176be092/build/components/pbkdf1.js)

File: AppInspector/rules/default/frameworks/rust.json [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    "description": "Development Framework: Rocket",
+    "applies_to": [ "rust" ],
+    "tags": [ "Framework.Development.Rocket" ],
+    "severity": "moderate",
+    "patterns": [
+      {
+        "pattern": "extern crate rocket;|use rocket",
+        "type": "regex",
+        "scopes": [ "code" ],
+        "modifiers": [ "i" ]
```
Comment (scovetta) [inline AppInspector/rules/default/frameworks/rust.json:14]: I don't think you need these `i` modifiers -- these should be case sensitive.

Comment (guyacosta) [reply AppInspector/rules/default/frameworks/rust.json:14]: copy paste -will do

File: AppInspector/rules/default/general/code_metrics.json [hunk 3]
```diff
@@ -7,14 +7,33 @@
...
+  },
+  {
+    "name": "Metric: Struct Definition",
+    "id": "AI025710",
+    "description": "Metric: Struct Defined",
+    "tags": [
+      "Metric.Code.Class.Defined"
+    ],
+    "severity": "moderate",
+    "applies_to": [ "c", "cpp", "rust" ],
```
Comment (scovetta) [inline AppInspector/rules/default/general/code_metrics.json:29]: I think structs apply to Java and C# too (I'm sure more, but those off the top of my head.)

Comment (guyacosta) [reply AppInspector/rules/default/general/code_metrics.json:29]: "Java doesn't have structs like C or C++, but you can use Java classes and treat them like a struct." Added objective-c, c#

**Comment by guyacosta** [comment]
Interesting.  Just when you think a change is benign to unit testing.  I'll investigate and correct now.

File: AppInspector/rules/default/components/load_dll.json [hunk 4]
```diff
@@ -41,6 +41,27 @@
...
+    "applies_to": [ "rust" ],
+    "tags": [
+      "Component.Executable.Microsoft.DLL"
+    ],
+    "severity": "moderate",
+    "patterns": [
+      {
+        "pattern": "LoadLibrary|libloading|use std::unstable::dynamic_lib::DynamicLibrary",
+        "type": "regex",
+        "modifiers": [ "i" ],
```
Comment (gfs) [inline AppInspector/rules/default/components/load_dll.json:57]: Why case insensitive?

File: AppInspector/rules/default/frameworks/rust.json [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  {
+    "name": "Development Framework: Rocket",
+    "id": "AI025610",
+    "description": "Development Framework: Rocket",
+    "applies_to": [ "rust" ],
+    "tags": [ "Framework.Development.Rocket" ],
+    "severity": "moderate",
+    "patterns": [
+      {
+        "pattern": "extern crate rocket;|use rocket",
```
Comment (gfs) [inline AppInspector/rules/default/frameworks/rust.json:11]: why does one of these have a semi colon?

File: AppInspector/rules/default/frameworks/rust.json [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    "description": "Development Framework: ActixWeb",
+    "applies_to": [ "rust" ],
+    "tags": [ "Framework.Development.ActixWeb" ],
+    "severity": "moderate",
+    "patterns": [
+      {
+        "pattern": "use actix_web",
+        "type": "regex",
+        "scopes": [ "code" ]
+    ]
```
Comment (gfs) [inline AppInspector/rules/default/frameworks/rust.json:29]: Missing ending brace

File: AppInspector/rules/default/networkcomms/outbound_network.json [hunk 7]
```diff
@@ -136,6 +136,30 @@
...
+    ],
+    "severity": "moderate",
+    "patterns": [
+      {
+        "pattern": "TcpStream::connect|HttpConnection",
+        "type": "regex",
+        "scopes": [
+          "code"
+        ],
+        "modifiers": [ "i" ],
```
Comment (gfs) [inline AppInspector/rules/default/networkcomms/outbound_network.json:157]: Should this be case insensitive?

File: AppInspector/rules/default/os/dynamic_execution.json [hunk 8]
```diff
@@ -103,6 +104,29 @@
...
+    "tags": [
+      "OS.Process.DynamicExecution"
+    ],
+    "severity": "moderate",
+    "patterns": [
+      {
+        "pattern": "use std::process::Command;|Command::new",
+        "type": "regex",
+        "scopes": [ "code" ],
+        "modifiers": [ "i" ],
```
Comment (gfs) [inline AppInspector/rules/default/os/dynamic_execution.json:124]: Case insensitive intended?

File: AppInspector/rules/default/os/system_registry.json [hunk 9]
```diff
@@ -42,6 +42,43 @@
...
+    ],
+    "conditions": [
+      {
+        "pattern": {
+          "pattern": "HKEY_",
+          "type": "regex-word",
+          "scopes": [
+            "code"
+          ],
+          "modifiers": [ "i" ],
```
Comment (gfs) [inline AppInspector/rules/default/os/system_registry.json:73]: isn't HKEY case sensitive?

File: AppInspector/rules/default/os/system_registry.json [hunk 10]
```diff
@@ -42,6 +42,43 @@
...
+    "tags": [
+      "OS.SystemRegistry.Write"
+    ],
+    "applies_to": [
+      "csharp"
+    ],
+    "severity": "moderate",
+    "patterns": [
+      {
+        "pattern": "\\.create_subkey\\(|\\.set_value\\(",
```
Comment (gfs) [inline AppInspector/rules/default/os/system_registry.json:58]: I think the confiedence on set value is definitely lower than create subkey

Comment (gfs) [reply AppInspector/rules/default/os/system_registry.json:58]: Also these are not the correct function names for C#

File: AppInspector/rules/default/frameworks/rust.json (hunk 6 above)
Comment (guyacosta) [reply AppInspector/rules/default/frameworks/rust.json:29]: Yeah that's the one found also.  I was just having issues with VS recognizing the file change to upload again but resolved now and will review remaining comments before pushing.

File: AppInspector/rules/default/frameworks/rust.json (hunk 5 above)
Comment (guyacosta) [reply AppInspector/rules/default/frameworks/rust.json:11]: No reason for the diff other than copy paste.  A semicolon would make it more accurate if there was a chance we would see one without but no diff other than extra character to have to match.  I'll remove.

File: AppInspector/rules/default/os/system_registry.json (hunk 10 above)
Comment (guyacosta) [reply AppInspector/rules/default/os/system_registry.json:58]: No they aren't and can't find what language I found those in but there's a lot of them and easy to get a couple mixed though good Issue record so far.  I've corrected.  Thanks.

**Comment by guyacosta** [comment]
I think this one is ready.

File: AppInspector/rules/default/os/system_registry.json [hunk 11]
```diff
@@ -190,23 +226,39 @@
...
+    ],
+    "conditions": [
+      {
+        "pattern": {
+          "pattern": "HKEY_",
+          "type": "regex-word",
+          "scopes": [
+            "code"
+          ],
+          "modifiers": [ "i" ],
```
Comment (gfs) [inline AppInspector/rules/default/os/system_registry.json:255]: I thought you changed this to be case insensitive

Comment (guyacosta) [reply AppInspector/rules/default/os/system_registry.json:255]: I did and missed one in a condition section at the end.  Fixing now.

Comment (guyacosta) [reply AppInspector/rules/default/os/system_registry.json]: Reviewed the file and others have exact api names and no need for it there also.  I had focused on the comment earlier for HKEY too narrowly but have removed then elsewhere also where not needed in that file.  Often a copy paste to add a new rule etc.  Sometimes I don't notice the i.

**Comment by guyacosta** [comment]
I've submitted a separate issue #253 to continue looking into possible use of insensitivity in patterns.  Let's button this one up just related to Rust changes.

**Comment by guyacosta** [comment]
btw I scanned the rust language https://github.com/rust-lang/rust among others for testing.

**Pull request merged**